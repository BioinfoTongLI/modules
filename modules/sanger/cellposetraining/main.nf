process CELLPOSETRAINING {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "docker.io/biocontainers/cellpose:3.1.0_cv1":
        "docker.io/biocontainers/cellpose:3.1.0_cv1" }"

    input:
    tuple val(meta), path(dir, stageAs: 'train'), path(test_dir, stageAs: 'test')
    val(model)
    val(learning_rate)
    val(weight_decay)
    val(n_epochs)

    output:
    tuple val(meta), path("test/models/${prefix}_retrained"), emit: retrained_model
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    def original_model = model? "${model}" : "cyto3"
    """
    python3 -m cellpose --train  \\
        --dir "./train" \\
        --test_dir "./test" \\
        --pretrained_model ${original_model} \\
        --learning_rate ${learning_rate} \\
        --weight_decay ${weight_decay} \\
        --n_epochs ${n_epochs} \\
        --model_name_out ${prefix}_retrained \\
        --SGD 1 \\
        --use_gpu \\
        --mask_filter '_seg.npy' \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cellposetraining: \$(python3 -m cellpose --version)
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch test/models/${prefix}_retrained

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cellposetraining: \$(python3 -m cellpose --version)
    END_VERSIONS
    """
}
