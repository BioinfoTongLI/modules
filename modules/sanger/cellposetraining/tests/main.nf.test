nextflow_process {

    name "Test Process CELLPOSETRAINING"
    script "../main.nf"
    process "CELLPOSETRAINING"
    config "./nextflow.config"

    tag "modules"
    tag "modules_sanger"
    tag "cellposetraining"

    test("cellpose - npy") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file("/lustre/scratch127/cellgen/cellgeni/tl10/temp/xenium_morphology_mip/", checkIfExists: true),
                    file("/lustre/scratch127/cellgen/cellgeni/tl10/temp/xenium_morphology_mip/", checkIfExists: true),
                ]
                input[1] = "cyto3"
                input[2] = 0.1
                input[3] = 0.001
                input[4] = 30
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("cellpose - npy - stub") {

        options "-stub"
        
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file("/lustre/scratch127/cellgen/cellgeni/tl10/temp/xenium_morphology_mip/", checkIfExists: true),
                    file("/lustre/scratch127/cellgen/cellgeni/tl10/temp/xenium_morphology_mip/", checkIfExists: true),
                ]
                input[1] = "cyto3"
                input[2] = 0.1
                input[3] = 0.001
                input[4] = 30
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}
