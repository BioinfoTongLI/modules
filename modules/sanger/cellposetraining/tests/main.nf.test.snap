{
    "cellpose - npy": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_retrained:md5,0fdfb40f1b103d5128cdd003ca0875fa"
                    ]
                ],
                "1": [
                    "versions.yml:md5,463af6bfec4237cdc66a5078a08ba23f"
                ],
                "retrained_model": [
                    [
                        {
                            "id": "test"
                        },
                        "test_retrained:md5,0fdfb40f1b103d5128cdd003ca0875fa"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,463af6bfec4237cdc66a5078a08ba23f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-11T00:16:53.544836332"
    },
    "cellpose - npy - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_retrained:md5,0fdfb40f1b103d5128cdd003ca0875fa"
                    ]
                ],
                "1": [
                    "versions.yml:md5,463af6bfec4237cdc66a5078a08ba23f"
                ],
                "retrained_model": [
                    [
                        {
                            "id": "test"
                        },
                        "test_retrained:md5,0fdfb40f1b103d5128cdd003ca0875fa"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,463af6bfec4237cdc66a5078a08ba23f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-11T00:17:06.655012621"
    }
}